### PR TITLE
Image sharer task 6

### LINKED ISSUE #6: Image Tags
#### As a user I want to add tags to images.

__Story__:

Our users are loving the ability to add images to our service. However, now
that there are more than a handful of images, it is apparent that we need some
way to filter images by certain categories. In order to do that, we first need
to label images in some way. It is time to add the ability for our users to tag
images when they add them.

__Acceptance criteria__:
- [x] When I add a new image, there is a form field to add tags.
- [x] After I've saved an image with tags, I see those tags displayed with the
  associated image.
- [x] My application uses the acts-as-taggable-on gem to provide the tagging
  functionality.

__Reference material__:
- https://github.com/mbleigh/acts-as-taggable-on

__Note__: You should allow images to have no tags; this sets the stage for a later issue.

__Dependencies__:
- Image Index

### PR DESCRIPTION
**Pull request by revatidamle**
Closes #6 and #7

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [ ] I have replaced `X` above with the associated issue number.
- [ ] I have changed the issue's label to "pull request".
- [ ] I have assigned the designated reviewer to this PR.
- [ ] I have set the appropriate label on this PR.
- [ ] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: app/assets/stylesheets/main.scss [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  width: 100%;
+  text-align: center;
+}
+
+#logo {
+  background-image: url('assets/icon.png');
+  display: block;
+}
+
+.container-fluid{
```
Comment (grahamotte) [inline app/assets/stylesheets/main.scss]: Doesn't bootstrap give us this class already?

File: app/views/images/_card.html.erb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<div class="card" id=<%= 'id_' + image[:title]%>">
+  <img class="card-img-top" id="thumbnail" src="<%=image[:link]%>" />
+    <div class="card-body">
+      <h5 class="card-title"><%=image[:title]%></h5>
+      <p class="card-text">
+        <%image.tag_list.each do |tag| %>
+          <%= link_to tag , tagged_path(tag: tag)%>
+          <% end %>
```
Comment (grahamotte) [inline app/views/images/_card.html.erb]: alignment

File: app/views/images/index.html.erb [hunk 3]
```diff
@@ -1,22 +1,18 @@
...
-<table>
-  <tr>
-    <th>Title</th>
-    <th>Image</th>
-  </tr>
 <nav class="navbar navbar-light bg-light">
-  <div style="text-align:center color: white">Image sharer app</div><br>
-  <form class="form-inline">
-    <div style="padding-left: 10px;">
+  <div>Image sharer app</div><br>
```
Comment (grahamotte) [inline app/views/images/index.html.erb]: It's good practice to end void elements with a `/` like `<br />`. Also the `br` should prob be on the next line

File: app/views/images/tagged.html.erb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+<div class="container-fluid">
+  <% @image.each do |image|%>
+      <%= render partial: 'images/card', locals: { image: image } %>
```
Comment (grahamotte) [inline app/views/images/tagged.html.erb]: extra indent

File: app/views/images/index.html.erb [hunk 5]
```diff
@@ -1,22 +1,18 @@
...
+<div class="container-fluid">
   <% @all_images.each do |image|%>
-      <tr id= <%= 'id_' + image[:title] %>>
-        <td><%=image.title %></td>
-        <td><%= image_tag  image.link, width: 400, crop: "pad"%></td>
-      </tr>
-  <%end %>
-</table>
+       <%= render partial: 'images/card', locals: { image: image } %>
+<%end%>
```
Comment (grahamotte) [inline app/views/images/index.html.erb]: some alignment issues here and the line above

File: db/migrate/20190305010613_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  class ActsAsTaggableOnMigration < ActiveRecord::Migration; end
+end
+ActsAsTaggableOnMigration.class_eval do
+  def self.up
+    create_table :tags do |t|
+      t.string :name
+    end
+
+    create_table :taggings do |t|
+      t.references :tag
```
Comment (grahamotte) [inline db/migrate/20190305010613_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: I think at least some of these fields should be `t.references :whatev, null: false`. We wouldn't want to get in a case where we are saving taggings with no tag :)

File: app/views/images/_card.html.erb (hunk 2 above)
Comment (revatidamle) [reply app/views/images/_card.html.erb]: not sure why the rubocop doesn't fix these things :-o

Comment (grahamotte) [reply app/views/images/_card.html.erb]: `rubocop --auto-correct` should, though I'm not sure it handles `erb`s

File: app/views/images/new.html.erb [hunk 7]
```diff
@@ -1,5 +1,6 @@
 <%= simple_form_for @image, defaults: { input_html: { class: 'form-control' } } do |i| %>
   <%= i.input :title %>
   <%= i.input :link %>
+    <%= i.input :tag_list %>
```
Comment (joelbandi) [inline app/views/images/new.html.erb]: indentation

File: app/controllers/images_controller.rb [hunk 8]
```diff
@@ -17,10 +17,17 @@ def create
     end
   end
 
+  def tagged
```
Comment (joelbandi) [inline app/controllers/images_controller.rb]: This seems like a weird URL pattern for resources to me, In order to make it completely restful 
if we visit `www.test.com/images` -> Display all images with tags too by default
if we visit `www.test.com/images?tag=some_tag` -> Display all images with tags `some_tag`

but currently we do`www.test.com/tagges?tag=some_tag` -> to display all images with tags `some_tag`

File: app/views/images/tagged.html.erb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+<div class="image_grid">
```
Comment (joelbandi) [inline app/views/images/tagged.html.erb]: Seems weird to have two view files almost identical

File: app/controllers/images_controller.rb (hunk 8 above)
Comment (revatidamle) [reply app/controllers/images_controller.rb]: Makes sense. Changed it! 👍

File: app/assets/stylesheets/main.scss (hunk 1 above)
Comment (revatidamle) [reply app/assets/stylesheets/main.scss]: Thats right. I had a little bit of confusion on using the bootstrap classes. Have made the necessary changes.

File: app/controllers/images_controller.rb [hunk 10]
```diff
@@ -1,6 +1,10 @@
 class ImagesController < ApplicationController
   def index
-    @all_images = Image.all.reverse
+    @image = if params[:tag].present?
+               Image.tagged_with(params[:tag])
```
Comment (joelbandi) [inline app/controllers/images_controller.rb]: Can you try looking into scopes for example `default_scope { order("created_at DESC") }`

File: app/views/images/_card.html.erb [hunk 11]
```diff
@@ -2,6 +2,11 @@
   <img class="card__thumbnail" src="<%=image[:link]%>"/>
   <div class="card__body">
     <h5 class="card__title"><%=image[:title]%></h5>
+    <p class="card-text">
+      <%image.tag_list.each do |tag| %>
+          <%= link_to tag , images_path(tag: tag)%>
```
Comment (joelbandi) [inline app/views/images/_card.html.erb]: 2 spaces indentation 😢

Comment (joelbandi) [reply app/views/images/_card.html.erb]: Also as an aside the spacing around `,` is usually `test_method :arg1, :arg2, :arg3`

Comment (revatidamle) [reply app/views/images/_card.html.erb]: I don't know why the indentations don't work properly in my rubymine >.<

File: app/controllers/images_controller.rb (hunk 10 above)
Comment (revatidamle) [reply app/controllers/images_controller.rb]: Thanks, new knowledge addition :)

**Pull request merged**